modules/modules.go

```diff
@@ -0,0 +1,279 @@
+// "foo.loo", but not "bar", "foo.bar.loo", etc. An empty scope
+// returns top-level modules, for example "foo" or "bar". Partial
+// scopes are not matched (i.e. scope "foo.ba" does not match
+// name "foo.bar").
+//
+// Because modules are registered to a map under the hood, the
+// returned slice will be sorted to keep it deterministic.
+func GetModules(scope string) []Module {
+	modulesMu.RLock()
+	defer modulesMu.RUnlock()
+
+	scopeParts := strings.Split(scope, ".")
+
+	// handle the special case of an empty scope, which
+	// should match only the top-level modules
+	if scope == "" {
+		scopeParts = []string{}
+	}
+
+	var mods []Module
+iterateModules:
+	for id, m := range modules {
+		modParts := strings.Split(id, ".")
+
+		// match only the next level of nesting
+		if len(modParts) != len(scopeParts)+1 {
+			continue
+		}
+
+		// specified parts must be exact matches
+		for i := range scopeParts {
+			if modParts[i] != scopeParts[i] {
+				continue iterateModules
+			}
+		}
+
+		mods = append(mods, m)
+	}
+
+	// make return value deterministic
+	sort.Slice(mods, func(i, j int) bool {
+		return mods[i].K6Module().ID < mods[j].K6Module().ID
+	})
+
+	return mods
+}
+
+// Modules returns the names of all registered modules
+// in ascending lexicographical order.
+func Modules() []string {
+	modulesMu.RLock()
+	defer modulesMu.RUnlock()
+
+	names := make([]string, 0, len(modules))
+	for name := range modules {
+		names = append(names, name)
+	}
+
+	sort.Strings(names)
+
+	return names
+}
+
+// Provisioner is implemented by modules which may need to perform
+// some additional "setup" steps immediately after being loaded.
+// Provisioning should be fast (imperceptible running time). If
+// any side-effects result in the execution of this function (e.g.
+// creating global state, any other allocations which require
+// garbage collection, opening files, starting goroutines etc.),
+// be sure to clean up properly by implementing the CleanerUpper
+// interface to avoid leaking resources.
+type Provisioner interface {
+	// Provision(Context) error
+	Provision() error
+}
+
+// Validator is implemented by modules which can verify that their
+// configurations are valid. This method will be called after
+// Provision() (if implemented). Validation should always be fast
+// (imperceptible running time) and an error must be returned if
+// the module's configuration is invalid.
+type Validator interface {
+	Validate() error
+}
+
+// CleanerUpper is implemented by modules which may have side-effects
+// such as opened files, spawned goroutines, or allocated some sort
+// of non-stack state when they were provisioned. This method should
+// deallocate/cleanup those resources to prevent memory leaks. Cleanup
+// should be fast and efficient. Cleanup should work even if Provision
+// returns an error, to allow cleaning up from partial provisionings.
+type CleanerUpper interface {
+	Cleanup() error
```

Comment:
These interfaces are unused right now, but they seem generally useful to have. Should they be hooked up?